### ISSUE TITLE
JVM crash

### ISSUE THREAD

**Issue by yanns** [drrb/java-rust-example#5]
Hi,

first of all, thanks for your project!

I just tried it on mac, and the JVM crash:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.github.drrb.javarust.GreetingsTest
Starting shouldGetAStructFromRustByValue(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringParameterFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
SUREFIRE-859: Hello from Rust, World
Starting shouldGetAStructFromRustByReference(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringFromJavaToRustAndReturnAnotherOne(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptAStructFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
Starting shouldGetAStructFromRustContainingAnArrayOfStructs(com.github.drrb.javarust.GreetingsTest)
#
# A fatal error has been detected by the Java Runtime Environment:
#
SUREFIRE-859: #  SIGSEGV (0xb) at pc=0x00007fff98c60152, pid=6353, tid=0x0000000000001703
#
# JRE version: Java(TM) SE Runtime Environment (8.0_92-b14) (build 1.8.0_92-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.92-b14 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libsystem_c.dylib+0x1152]  strlen+0x12
#
SUREFIRE-859: # Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/yannsimon/projects/rust/java-rust-example/hs_err_pid6353.log
#
SUREFIRE-859: # If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
/bin/sh: line 1:  6353 Abort trap: 6           /Library/Java/JavaVirtualMachines/jdk1.8.0_92.jdk/Contents/Home/jre/bin/java -Djna.nosys=true -jar /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefirebooter3340298246630967947.jar /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefire2630333911108497694tmp /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefire_08200565017308027989tmp

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.787 s
[INFO] Finished at: 2016-06-18T16:11:30+02:00
[INFO] Final Memory: 17M/211M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.18.1:test (default-test) on project java-rust-example: ExecutionException: java.lang.RuntimeException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /Users/yannsimon/projects/rust/java-rust-example && /Library/Java/JavaVirtualMachines/jdk1.8.0_92.jdk/Contents/Home/jre/bin/java -Djna.nosys=true -jar /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefirebooter3340298246630967947.jar /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefire2630333911108497694tmp /Users/yannsimon/projects/rust/java-rust-example/target/surefire/surefire_08200565017308027989tmp
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

If I can help debugging the issue, please just tell me what you need.

**Comment by Dushistov** [comment]
I got the same issue on amd64 Linux with:
# JRE version: Java(TM) SE Runtime Environment (7.0_80-b15) (build 1.7.0_80-b15)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.80-b11 mixed mode linux-amd64 compressed oops)
and rust 1.9

**Comment by twall** [comment]
Most likely your structure mapping is incorrect.

Show your struct mapping in native and in Java code for comparison.

**Comment by yanns** [comment]
the mappings are the one found in the code. I just run `mvn test`

**Comment by twall** [comment]
Presumably this works on some _other_ platforms; I'm not sure what would be
different about OS X or amd64 other than pointer or native long type sizing.

**Comment by twall** [comment]
I'm not that familiar with Rust, but the storage of pointers to string objects independently of the the original strings seems a little suspect.   Depending on how Rust manages references to strings, you might wind up with the JNA side trying to access memory that's no longer valid.

**Comment by zimy** [comment]
Have almost same issue with running mvn package on Fedora 24 on Intel i5-4200 with Rust 1.10 and OpenJDK 8:
```
 T E S T S
-------------------------------------------------------
Running com.github.drrb.javarust.GreetingsTest
Starting shouldGetAStructFromRustByValue(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringParameterFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
SUREFIRE-859: Hello from Rust, World
Starting shouldGetAStructFromRustByReference(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringFromJavaToRustAndReturnAnotherOne(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptAStructFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
Starting shouldGetAStructFromRustContainingAnArrayOfStructs(com.github.drrb.javarust.GreetingsTest)
Starting shouldGetAStringFromRustInACallback(com.github.drrb.javarust.GreetingsTest)
Starting shouldGetListOfStringsFromRustInACallback(com.github.drrb.javarust.GreetingsTest)
*** Error in `/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.101-1.b14.fc24.x86_64/jre/bin/java': free(): invalid pointer: 0x00007fd390347ad8 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x755fb)[0x7fd3992b45fb]
/lib64/libc.so.6(+0x7dd4a)[0x7fd3992bcd4a]
/lib64/libc.so.6(cfree+0x4c)[0x7fd3992c02ac]
/home/zimy/Documents/hot_projects/java-rust-example/target/classes/linux-x86-64/libgreetings.so(+0xb4dd5)[0x7fd3808f7dd5]
/home/zimy/Documents/hot_projects/java-rust-example/target/classes/linux-x86-64/libgreetings.so(+0xb5aa5)[0x7fd3808f8aa5]
/home/zimy/Documents/hot_projects/java-rust-example/target/classes/linux-x86-64/libgreetings.so(+0xb6a02)[0x7fd3808f9a02]
/home/zimy/Documents/hot_projects/java-rust-example/target/classes/linux-x86-64/libgreetings.so(dropGreeting+0x11)[0x7fd3808f99c1]
/tmp/jna-3738907/jna1189522310089792417.tmp(ffi_call_unix64+0x4c)[0x7fd380bf77a0]
/tmp/jna-3738907/jna1189522310089792417.tmp(ffi_call+0x20b)[0x7fd380bf729b]
/tmp/jna-3738907/jna1189522310089792417.tmp(+0x5e41)[0x7fd380beae41]
/tmp/jna-3738907/jna1189522310089792417.tmp(Java_com_sun_jna_Native_invokeVoid+0x22)[0x7fd380bed752]
[0x7fd381017754]
======= Memory map: ========
........
```

**Comment by bigkahuna1uk** [comment]
Happened for me just building straight out of the box on OS X El Capaitan with Java 1.8.121 and mvn 3.3.9

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.github.drrb.javarust.GreetingsTest
Starting shouldGetAStructFromRustByValue(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringParameterFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
SUREFIRE-859: Hello from Rust, World
Starting shouldGetAStructFromRustByReference(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptStringFromJavaToRustAndReturnAnotherOne(com.github.drrb.javarust.GreetingsTest)
Starting shouldAcceptAStructFromJavaToRust(com.github.drrb.javarust.GreetingsTest)
Starting shouldGetAStructFromRustContainingAnArrayOfStructs(com.github.drrb.javarust.GreetingsTest)
#
# A fatal error has been detected by the Java Runtime Environment:
#
SUREFIRE-859: #  SIGSEGV (0xb) at pc=0x00007fffa35bfb52, pid=33524, tid=0x0000000000001c03
#
# JRE version: Java(TM) SE Runtime Environment (8.0_121-b13) (build 1.8.0_121-b13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.121-b13 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libsystem_c.dylib+0x1b52]  strlen+0x12
#
# Core dump written. Default location: /cores/core or core.33524
#
# An error report file with more information is saved as:
# /Users/kellyc/development/github/java-rust-example/hs_err_pid33524.log
#
SUREFIRE-859: # If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
/bin/sh: line 1: 33524 Abort trap: 6           (core dumped) /Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home/jre/bin/java -Djna.nosys=true -jar /Users/kellyc/development/github/java-rust-example/target/surefire/surefirebooter894421574483473332.jar /Users/kellyc/development/github/java-rust-example/target/surefire/surefire570168086942519241tmp /Users/kellyc/development/github/java-rust-example/target/surefire/surefire_05668102102819219134tmp

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 31.660s
[INFO] Finished at: Wed Mar 15 10:43:48 GMT 2017
[INFO] Final Memory: 21M/981M
```

**Comment by drrb** [comment]
This has since been fixed (in 81ede5d6be4f195865b8d39e51631a812a2c6c4b), closing.
